### ISSUE TITLE
Assertion `successors.size() == 1' failed

### ISSUE THREAD

**Issue by wcventure** [mchalupa/dg#302]
Hi there, 
Could you help me to find out the problem? I use the slicer on seveal program, but I have triggered some assertion failed. Do I have any way to avoid this situation and continue the program slicing.

The first example (program grep.bc in Grep 2.5.4)
I upload the bc file, you can download it from here.
[grep.zip](https://github.com/mchalupa/dg/files/3882111/grep.zip)

When I runing `llvm-slicer -c malloc ./grep.bc`, I got assertion failed
```
/dg/include/dg/analysis/SubgraphNode.h:181: NodeT* dg::analysis::SubgraphNode<NodeT>::getSingleSuccessor() const [with NodeT = dg::analysis::pta::PSNode]: Assertion `successors.size() == 1' failed.
Aborted (core dumped)
```

Another example (program objdump.bc in binutils 2.28)
I upload the bc file, you can download it from here.
[objdump.zip](https://github.com/mchalupa/dg/files/3882112/objdump.zip)

When I runing `dg/tools/llvmdg-show objdump.bc`, I got assertion failed
```
dg/include/dg/analysis/SubgraphNode.h:181: NodeT* dg::analysis::SubgraphNode<NodeT>::getSingleSuccessor() const [with NodeT = dg::analysis::pta::PSNode]: Assertion `successors.size() == 1' failed.
Aborted (core dumped)
```

**Comment by mchalupa** [comment]
Hi,
this problem should be fixed in f4cbff52d4fd51f. You can cherry-pick it into your local branch (I will merge it with the rest of the svcomp19 branch after some testing).

**Comment by wcventure** [comment]
Hi mchalupa,

Excuse me, I try use f4cbff5 to extract the PDG for grep 2.5.4. Unfortunately, another assertion failed occur.

```
llvm-slicer: /dg/lib/llvm/ReadWriteGraph/LLVMReadWriteGraphBuilder.cpp:512: dg::dda::LLVMReadWriteGraphBuilder::Subgraph* dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(const llvm::Function*): Assertion `subg->entry && "No entry in the subgraph"' failed.
#0 0x00005565114acad7 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /home/wencheng/Research/CLib/llvm-6.0.1/lib/Support/Unix/Signals.inc:398:0
#1 0x00005565114acb6a PrintStackTraceSignalHandler(void*) /home/wencheng/Research/CLib/llvm-6.0.1/lib/Support/Unix/Signals.inc:462:0
#2 0x00005565114aad3f llvm::sys::RunSignalHandlers() /home/wencheng/Research/CLib/llvm-6.0.1/lib/Support/Signals.cpp:49:0
#3 0x00005565114ac343 SignalHandler(int) /home/wencheng/Research/CLib/llvm-6.0.1/lib/Support/Unix/Signals.inc:252:0
#4 0x00007f28e5cdd890 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12890)
#5 0x00007f28e4fd9e97 gsignal /build/glibc-OTsEL5/glibc-2.27/signal/../sysdeps/unix/sysv/linux/raise.c:51:0
#6 0x00007f28e4fdb801 abort /build/glibc-OTsEL5/glibc-2.27/stdlib/abort.c:81:0
#7 0x00007f28e4fcb39a __assert_fail_base /build/glibc-OTsEL5/glibc-2.27/assert/assert.c:89:0
#8 0x00007f28e4fcb412 (/lib/x86_64-linux-gnu/libc.so.6+0x30412)
#9 0x00007f28e69d1679 dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(llvm::Function const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c679)
#10 0x00007f28e69d19e5 dg::dda::LLVMReadWriteGraphBuilder::createCallToFunction(llvm::Function const*, llvm::CallInst const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c9e5)
#11 0x00007f28e69d3b1b dg::dda::LLVMReadWriteGraphBuilder::createCall(llvm::Instruction const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7eb1b)
#12 0x00007f28e69d1130 dg::dda::LLVMReadWriteGraphBuilder::buildBlockNodes(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c130)
#13 0x00007f28e69d1244 dg::dda::LLVMReadWriteGraphBuilder::buildBlock(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c244)
#14 0x00007f28e69d259c dg::dda::LLVMReadWriteGraphBuilder::buildFunction(llvm::Function const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7d59c)
#15 0x00007f28e69d160d dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(llvm::Function const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c60d)
#16 0x00007f28e69d19e5 dg::dda::LLVMReadWriteGraphBuilder::createCallToFunction(llvm::Function const*, llvm::CallInst const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c9e5)
#17 0x00007f28e69d3b1b dg::dda::LLVMReadWriteGraphBuilder::createCall(llvm::Instruction const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7eb1b)
#18 0x00007f28e69d1130 dg::dda::LLVMReadWriteGraphBuilder::buildBlockNodes(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c130)
#19 0x00007f28e69d1244 dg::dda::LLVMReadWriteGraphBuilder::buildBlock(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c244)
#20 0x00007f28e69d259c dg::dda::LLVMReadWriteGraphBuilder::buildFunction(llvm::Function const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7d59c)
#21 0x00007f28e69d160d dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(llvm::Function const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c60d)
#22 0x00007f28e69d19e5 dg::dda::LLVMReadWriteGraphBuilder::createCallToFunction(llvm::Function const*, llvm::CallInst const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c9e5)
#23 0x00007f28e69d3b1b dg::dda::LLVMReadWriteGraphBuilder::createCall(llvm::Instruction const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7eb1b)
#24 0x00007f28e69d1130 dg::dda::LLVMReadWriteGraphBuilder::buildBlockNodes(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c130)
#25 0x00007f28e69d1244 dg::dda::LLVMReadWriteGraphBuilder::buildBlock(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c244)
#26 0x00007f28e69d259c dg::dda::LLVMReadWriteGraphBuilder::buildFunction(llvm::Function const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7d59c)
#27 0x00007f28e69d160d dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(llvm::Function const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c60d)
#28 0x00007f28e69d19e5 dg::dda::LLVMReadWriteGraphBuilder::createCallToFunction(llvm::Function const*, llvm::CallInst const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c9e5)
#29 0x00007f28e69d3b1b dg::dda::LLVMReadWriteGraphBuilder::createCall(llvm::Instruction const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7eb1b)
#30 0x00007f28e69d1130 dg::dda::LLVMReadWriteGraphBuilder::buildBlockNodes(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c130)
#31 0x00007f28e69d1244 dg::dda::LLVMReadWriteGraphBuilder::buildBlock(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c244)
#32 0x00007f28e69d259c dg::dda::LLVMReadWriteGraphBuilder::buildFunction(llvm::Function const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7d59c)
#33 0x00007f28e69d160d dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(llvm::Function const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c60d)
#34 0x00007f28e69d19e5 dg::dda::LLVMReadWriteGraphBuilder::createCallToFunction(llvm::Function const*, llvm::CallInst const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c9e5)
#35 0x00007f28e69d3b1b dg::dda::LLVMReadWriteGraphBuilder::createCall(llvm::Instruction const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7eb1b)
#36 0x00007f28e69d1130 dg::dda::LLVMReadWriteGraphBuilder::buildBlockNodes(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c130)
#37 0x00007f28e69d1244 dg::dda::LLVMReadWriteGraphBuilder::buildBlock(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c244)
#38 0x00007f28e69d259c dg::dda::LLVMReadWriteGraphBuilder::buildFunction(llvm::Function const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7d59c)
#39 0x00007f28e69d160d dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(llvm::Function const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c60d)
#40 0x00007f28e69d19e5 dg::dda::LLVMReadWriteGraphBuilder::createCallToFunction(llvm::Function const*, llvm::CallInst const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c9e5)
#41 0x00007f28e69d3b1b dg::dda::LLVMReadWriteGraphBuilder::createCall(llvm::Instruction const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7eb1b)
#42 0x00007f28e69d1130 dg::dda::LLVMReadWriteGraphBuilder::buildBlockNodes(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c130)
#43 0x00007f28e69d1244 dg::dda::LLVMReadWriteGraphBuilder::buildBlock(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c244)
#44 0x00007f28e69d259c dg::dda::LLVMReadWriteGraphBuilder::buildFunction(llvm::Function const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7d59c)
#45 0x00007f28e69d160d dg::dda::LLVMReadWriteGraphBuilder::getOrCreateSubgraph(llvm::Function const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c60d)
#46 0x00007f28e69d1dfb dg::dda::LLVMReadWriteGraphBuilder::createCallToFunctions(std::vector<llvm::Function const*, std::allocator<llvm::Function const*> > const&, llvm::CallInst const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7cdfb)
#47 0x00007f28e69d3bfd dg::dda::LLVMReadWriteGraphBuilder::createCall(llvm::Instruction const*) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7ebfd)
#48 0x00007f28e69d1130 dg::dda::LLVMReadWriteGraphBuilder::buildBlockNodes(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c130)
#49 0x00007f28e69d1244 dg::dda::LLVMReadWriteGraphBuilder::buildBlock(dg::dda::LLVMReadWriteGraphBuilder::Subgraph&, llvm::BasicBlock const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7c244)
#50 0x00007f28e69d259c dg::dda::LLVMReadWriteGraphBuilder::buildFunction(llvm::Function const&) (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7d59c)
#51 0x00007f28e69d454c dg::dda::LLVMReadWriteGraphBuilder::build() (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x7f54c)
#52 0x00007f28e69ee870 dg::dda::LLVMDataDependenceAnalysis::createDDA() (/home/wencheng/Research/FuzzingProject/dg-new/lib/libdgllvmdda.so+0x99870)
#53 0x000055651134b103 dg::dda::LLVMDataDependenceAnalysis::buildGraph() (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3ca103)
#54 0x000055651134b179 dg::dda::LLVMDataDependenceAnalysis::run() (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3ca179)
#55 0x000055651134b332 dg::llvmdg::LLVMDependenceGraphBuilder::_runDataDependenceAnalysis() (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3ca332)
#56 0x000055651134bc1c dg::llvmdg::LLVMDependenceGraphBuilder::computeDependencies(std::unique_ptr<dg::LLVMDependenceGraph, std::default_delete<dg::LLVMDependenceGraph> >&&) (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3cac1c)
#57 0x000055651134ef3f Slicer::computeDependencies() (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3cdf3f)
#58 0x000055651134f186 Slicer::mark(std::set<dg::LLVMNode*, std::less<dg::LLVMNode*>, std::allocator<dg::LLVMNode*> >&) (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3ce186)
#59 0x0000556511341481 main (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3c0481)
#60 0x00007f28e4fbcb97 __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:344:0
#61 0x000055651134031a _start (/home/wencheng/Research/FuzzingProject/dg-new/tools/llvm-slicer+0x3bf31a)
Aborted (core dumped)
```
